Migrate to app router

mark-tate opened the pull request:
Yet to go's
1. Fix up tests
2. Update docs
3. Update standard generator

The site works with app router, just run
```
yarn serve
```

The goal of this PR was a surgical PR that does the minimum amount of refactoring to support server components and server functions

mark-tate committed:
upgrade esbuild to 0.17.19 to preserve directive `"use client"`for React Server Components support

mark-tate committed:
update Next to 13.4.4 to resolve issue with loading React Server Components

mark-tate committed:
remove context from store for rsc support

mark-tate committed:
convert site to use app router

mark-tate committed:
update `next-mdx/remote' to support rsc

mark-tate committed:
add `'use client'` directive to components for app router support

mark-tate committed:
add `'use client'` directive to site components for app router support

mark-tate committed:
refactor layouts to work with app router

mark-tate committed:
remove markdown components into `@jpmorganchase/mosaic-components-client` package

mark-tate committed:
move mdx export from `@jpmorganchase/mosaic-site-components` to `@jpmorganchase/mosaic-components-server`

mark-tate committed:
new packages for client-side MDX components called `@jpmorganchase/mosaic-components-client`

mark-tate committed:
new `@jpmorganchase/mosaic-components-server` package to support Remote Server Components and App Router in site

mark-tate committed:
new package for icons assets `@jpmorganchase/mosaic-icons`

previously icons were exported from theme as components but theme is now CSS only

mark-tate committed:
move ThemeProvider to `@jpmorganchase/site-components` to remove dependency on store in components package

mark-tate commented on packages/components/package.json:
```diff
@@ -40,7 +40,7 @@
     "typescript": "^4.8.3"
   },
   "dependencies": {
-    "@jpmorganchase/mosaic-store": "^0.1.0-beta.36",
```
NOTE: Removed shady dependency on store by moving ThemeProvider to site-components

mark-tate commented on packages/components/src/Markdown/index.tsx:
NOTE: Markdown has moved to it's own package `mdx-components-client`

mark-tate commented on packages/icons/package.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "@jpmorganchase/mosaic-icons",
```
New package for icons, as previously icons were exported from theme package but we want the theme package to be CSS only (otherwise, the CSS build thinks they are server components)

mark-tate commented on packages/layouts/src/layouts/Edit/index.tsx:
```diff
@@ -1,35 +1,26 @@
-  const router = useRouter();
   const { pageState, stopEditing } = useContentEditor();
-
-  useEffect(() => {
```
Now a Server component so we removed hook and replaced with <NavigationEvents/>

mark-tate commented on packages/site-components/package.json:
```diff
@@ -60,8 +61,8 @@
-    "next": "^13.4.1",
-    "next-mdx-remote": "^4.2.1",
+    "next": "^13.4.4",
+    "next-mdx-remote": "^4.4.1",
```
v4.4.1 supports RSCs

mark-tate committed:
simplify the configuration of site by moving providers to site components

mark-tate committed:
remove un-necessary 'use client' directives from site-components

DavieReid commented on packages/site/src/app/layout.tsx:
```diff
@@ -0,0 +1,16 @@
+import Head from 'next/head';
```
I think there is a metadata API now

https://nextjs.org/blog/next-13-3#file-based-metadata-api

mark-tate replied:
I've not looked at Search yet but GTK

mark-tate committed:
remove open-api and lab components from default component packages

if you require these components, add them into your configuration

mark-tate committed:
add `@jpmorganchase/site-loaders` which add loaders to site for supported modes

mark-tate committed:
rebase on beta.38

mark-tate committed:
move favicon from static images to app dir and don't reuse for logo

instead we create a `logo.png` that makes it obvious that someone can replace with
a logo and that favicon.ico is separate

mark-tate committed:
move next-auth to page router to avoid clash with dynamic routes in app router

When running locally this causes api urls to be passed to `app/[...slug]/page`

mark-tate committed:
refactor MDX loader to support static site exports

- renamed `@mosaic-site-loaders` to `@mosaic-site-mdx-loader`
- added support for static export of sites with `yarn export:static`
- added serve command for static site with `yarn serve:static`
- refactored page loader to support redirects
- static site exports use GENERATE_STATIC_PARAMS_URL env var to define the FS Url

mark-tate committed:
add 404 and 500 support

Create custom pages for the unhappy path using Next JS file conventions (`error.tsx`, `not-found.tsx`)

mark-tate committed:
upgrade Next to 13.4.9 with improved build times

removed webpack change previously created by React Live config

mark-tate committed:
add un-optimized support for SearchInput

mark-tate committed:
remove duplicate local env var for NEXTAUTH_URL

mark-tate committed:
remove redundant `@jpmorganchase/mosaic-site-middleware` replaced by Next JS 13 features such as app router and static export

mark-tate committed:
remove snaphot support
- removed commands from package.json and cli
- removed snapshot dir
- remove snapshot mode types

mark-tate committed:
remove `transpilePackages` from `next.config.ts`

mark-tate committed:
update Next to 13.4.10

mark-tate committed:
rebase on beta.39

mark-tate committed:
generate static pages for site for creating a build

mark-tate committed:
ensure static sites 404 pages are served when file not found

mark-tate committed:
rebase on beta.40

mark-tate commented:
rebased on beta.47 and created a new branch, in case there are issues

closing this one
